### PR TITLE
Mark flow elephant/v6

### PR DESCRIPTION
**Pull request by inashivb**
Link to ticket: https://redmine.openinfosecfoundation.org/issues/5647

SV_BRANCH=https://github.com/OISF/suricata-verify/pull/2395

Previous PR: https://github.com/OISF/suricata/pull/12867

Changes since v1:
- more tests and edge case fixups
- doc updated
- s-v test
- rebased on top of latest master
- schema update
- counter + logging added

### CODE REVIEW

File: src/flow.c [hunk 1]
```diff
@@ -344,6 +345,24 @@ static inline void FlowUpdateTtlTC(Flow *f, uint8_t ttl)
     f->max_ttl_toclient = MAX(f->max_ttl_toclient, ttl);
 }
 
+static inline void FlowUpdateFlowRate(
+        ThreadVars *tv, DecodeThreadVars *dtv, Flow *f, const Packet *p, int dir)
+{
+    if (FlowRateStorageEnabled()) {
+        FlowRateStore *frs = FlowGetStorageById(f, FlowRateGetStorageID());
```
Comment (victorjulien) [inline src/flow.c:352]: should we check `f->flags & FLOW_IS_ELEPHANT` before the call to FlowGetStorageById?

File: src/util-flow-rate.c [hunk 2]
```diff
@@ -0,0 +1,626 @@
...
+    }
+
+    SCFree(frs);
+    total_free += sizeof(*frs);
+    (void)SC_ATOMIC_SUB(flow_memuse, total_free);
+}
+
+void FlowRateRegisterFlowStorage(void)
+{
+    ConfNode *root = ConfGetNode("flow");
```
Comment (victorjulien) [inline src/util-flow-rate.c:61]: these have a SC prefix now in master?

File: src/util-flow-rate.c [hunk 3]
```diff
@@ -0,0 +1,626 @@
...
+    }
+    SCTime_t interval = SCTIME_INITIALIZER;
+    uint16_t secs = 0;
+    if ((StringParseUint16(&secs, 10, 0, val) < 0) || (secs == 0)) {
+        FatalError("Invalid value for flow tracking interval");
+    }
+    flow_rate_config.interval = SCTIME_ADD_SECS(interval, secs);
+
+    g_flowrate_storage_id = FlowStorageRegister(
+            "flowrate", sizeof(void *), NULL, (void (*)(void *))FlowRateStoreFree);
```
Comment (victorjulien) [inline src/util-flow-rate.c:93]: can we avoid this cast?

File: src/util-flow-rate.c [hunk 4]
```diff
@@ -0,0 +1,626 @@
...
+    FlowRateStore *frs = NULL;
+    size_t total_memuse = 0;
+    size_t expected_memuse = (2 * flow_rate_config.interval.secs * sizeof(uint64_t)) + sizeof(*frs);
+
+    if (!FLOW_CHECK_MEMCAP(expected_memuse)) {
+        return NULL;
+    }
+    frs = SCCalloc(1, sizeof(*frs));
+    if (unlikely(frs == NULL)) {
+        SCLogError("Unable to allocate FlowRateStore memory");
```
Comment (victorjulien) [inline src/util-flow-rate.c:112]: this is called from the packet path, right? No errors or other output please

File: src/util-flow-rate.c [hunk 5]
```diff
@@ -0,0 +1,626 @@
...
+
+    total_memuse += sizeof(*frs);
+    for (int i = 0; i < 2; i++) {
+        frs->dir[i].size = flow_rate_config.interval.secs;
+        frs->dir[i].buf = SCCalloc(frs->dir[i].size, sizeof(uint64_t));
+        if (unlikely(frs->dir[i].buf == NULL)) {
+            FlowRateStoreFree(frs);
+            return NULL;
+        }
+        frs->dir[i].sum = 0;
```
Comment (victorjulien) [inline src/util-flow-rate.c:124]: nit: not needed to 0 these as SCCalloc did that already

File: src/util-flow-rate.c [hunk 6]
```diff
@@ -0,0 +1,626 @@
...
+        FlowRateStore *frs, uint16_t start, uint16_t end, int direction)
+{
+    for (uint16_t i = start; i <= end; i++) {
+        uint64_t byte_count_at_i = frs->dir[direction].buf[i];
+        frs->dir[direction].buf[i] = 0;
+        frs->dir[direction].sum -= byte_count_at_i;
+    }
+}
+
+void FlowRateStoreUpdate(FlowRateStore *frs, SCTime_t p_ts, uint32_t pkt_len, int direction)
```
Comment (victorjulien) [inline src/util-flow-rate.c:151]: style: I think this function is a bit long, can the 2 main branches be put into small (inline) helper funcs?

File: src/util-flow-rate.c [hunk 7]
```diff
@@ -0,0 +1,626 @@
...
+        frs->dir[direction].buf[0] = pkt_len;
+        /* Overwrite the sum calculated so far */
+        frs->dir[direction].sum = pkt_len;
+    }
+    SCLogDebug("Byte sum for dir %d: %" PRIu64, direction, frs->dir[direction].sum);
+    /* In any case, update the last seen timestamp */
+    frs->dir[direction].last_ts = p_ts;
+}
+
+bool FlowRateIsExceeding(ThreadVars *tv, DecodeThreadVars *dtv, FlowRateStore *frs, int direction)
```
Comment (victorjulien) [inline src/util-flow-rate.c:220]: feels a bit odd to update a counter here, seems more like something the caller should do

**Review by victorjulien — CHANGES_REQUESTED** [review]
Nice work. See some comments inline. Think we're very close.

File: src/flow.c (hunk 1 above)
Comment (inashivb) [reply src/flow.c:352]: Indeed that's much better

**Comment by suricata-qa** [comment]
ERROR:
 ERROR: buid failure for build_asan QA build

ERROR: QA failed on build_asan.
  

Pipeline 25484
